File: php/src/refactoring/golden-master/ReceiptPrinter.php
<?php

namespace Refactoring\GoldenMaster;

class OrderItem
{
    public function __construct(
        public string $sku,
        public string $description,
        public float $unitPrice,
        public int $quantity,
        public ?string $category = null
    ) {}
}

class Order
{
    public function __construct(
        public string $id,
        public string $customerName,
        public array $items
    ) {}
}

class ReceiptPrinter
{
    // Do not change this function at the beginning of the exercise; first create the Golden Master.
    public function print(Order $order): string
    {
        $now = $this->getCurrentDate();

        $header = sprintf("Recibo %s - %s", $order->id, $now->format('m/d/y H:i:s'));

        $subtotal = 0;
        $lines = [];
        $index = 1;

        foreach ($order->items as $item) {
            $lineTotal = $this->roundMoney($item->unitPrice * $item->quantity);
            $subtotal = $this->roundMoney($subtotal + $lineTotal);
            $lines[] = sprintf(
                "%d. %s (%s) x%d = $%.2f",
                $index++,
                $item->description,
                $item->sku,
                $item->quantity,
                $lineTotal
            );
        }

        $luckyDiscountPct = $this->discount();
        $luckyDiscount = $this->roundMoney($subtotal * $luckyDiscountPct);

        $taxableGeneral = 0.0;
        $foodTax = 0.0;

        foreach ($order->items as $item) {
            if ($item->category !== 'books' && $item->category !== 'food') {
                $taxableGeneral += $item->unitPrice * $item->quantity;
            }
            if ($item->category === 'food') {
                $foodTax += $item->unitPrice * $item->quantity * 0.03;
            }
        }

        $generalTax = $taxableGeneral * 0.07;
        $taxes = $this->roundMoney($generalTax + $foodTax);

        $total = $this->roundMoney($subtotal - $luckyDiscount + $taxes);

        $discountLine = $luckyDiscount > 0
            ? sprintf("Descuento de la suerte: -$%.2f (%.2f%%)", $luckyDiscount, $luckyDiscountPct * 100)
            : "Descuento de la suerte: $0.00 (0.00%)";

        $summary = [
            sprintf("Subtotal: $%.2f", $subtotal),
            $discountLine,
            sprintf("Impuestos: $%.2f", $taxes),
            sprintf("TOTAL: $%.2f", $total)
        ];

        $parts = array_merge([$header], $lines, ['---'], $summary);

        return implode("\n", $parts);
    }

    protected function discount(): float
    {
        $luckyDiscountPct = 0.0;
        if (mt_rand() / mt_getrandmax() < 0.1) {
            $luckyDiscountPct = (mt_rand() / mt_getrandmax()) * 0.05;
        }
        return $luckyDiscountPct;
    }

    protected function getCurrentDate(): \DateTime
    {
        return new \DateTime();
    }

    private function roundMoney(float $n): float
    {
        return round($n * 100) / 100;
    }
}

function generateOrder(string $id, string $customerName, int $numItems, int $quantity): Order
{
    $products = [
        ['sku' => 'BK-001', 'description' => 'Libro: Clean Code', 'unitPrice' => 30, 'category' => 'books'],
        ['sku' => 'FD-010', 'description' => 'Café en grano 1kg', 'unitPrice' => 12.5, 'category' => 'food'],
        ['sku' => 'GN-777', 'description' => 'Cuaderno A5', 'unitPrice' => 5.2, 'category' => 'general'],
        ['sku' => 'GN-123', 'description' => 'Bolígrafos (pack 10)', 'unitPrice' => 3.9, 'category' => 'general'],
        ['sku' => 'FD-222', 'description' => 'Té verde 200g', 'unitPrice' => 6.75, 'category' => 'food'],
    ];

    $items = [];
    for ($i = 0; $i < $numItems && $i < count($products); $i++) {
        $p = $products[$i];
        $items[] = new OrderItem(
            $p['sku'],
            $p['description'],
            $p['unitPrice'],
            $quantity,
            $p['category']
        );
    }

    return new Order($id, $customerName, $items);
}

File: java/src/main/java/com/refactoring/refactoring/ReceiptPrinter.java
package com.refactoring.refactoring;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Random;

/**
 * Prints receipts for orders.
 * Do not change this class at the beginning of the exercise; first create the Golden Master.
 */
public class ReceiptPrinter {
    private final Random random;

    public ReceiptPrinter() {
        this.random = new Random();
    }

    // Constructor for testing with seeded random
    public ReceiptPrinter(Random random) {
        this.random = random;
    }

    public String print(Order order) {
        Date now = getCurrentDate();
        SimpleDateFormat dateFormat = new SimpleDateFormat("MM/dd/yy HH:mm:ss");

        String header = String.format("Recibo %s - %s", order.getId(), dateFormat.format(now));

        double subtotal = 0.0;
        List<String> lines = new ArrayList<>();
        int index = 1;
        for (OrderItem item : order.getItems()) {
            double lineTotal = roundMoney(item.getUnitPrice() * item.getQuantity());
            subtotal = roundMoney(subtotal + lineTotal);
            lines.add(String.format("%d. %s (%s) x%d = $%.2f",
                    index++, item.getDescription(), item.getSku(), item.getQuantity(), lineTotal));
        }

        double luckyDiscountPct = discount();
        double luckyDiscount = roundMoney(subtotal * luckyDiscountPct);

        double taxableGeneral = 0.0;
        double foodTax = 0.0;
        for (OrderItem item : order.getItems()) {
            if (!"books".equals(item.getCategory()) && !"food".equals(item.getCategory())) {
                taxableGeneral += item.getUnitPrice() * item.getQuantity();
            }
            if ("food".equals(item.getCategory())) {
                foodTax += item.getUnitPrice() * item.getQuantity() * 0.03;
            }
        }
        double generalTax = taxableGeneral * 0.07;
        double taxes = roundMoney(generalTax + foodTax);

        double total = roundMoney(subtotal - luckyDiscount + taxes);

        String discountLine;
        if (luckyDiscount > 0) {
            discountLine = String.format("Descuento de la suerte: -$%.2f (%.2f%%)",
                    luckyDiscount, luckyDiscountPct * 100);
        } else {
            discountLine = "Descuento de la suerte: $0.00 (0.00%)";
        }

        List<String> summary = List.of(
                String.format("Subtotal: $%.2f", subtotal),
                discountLine,
                String.format("Impuestos: $%.2f", taxes),
                String.format("TOTAL: $%.2f", total)
        );

        List<String> parts = new ArrayList<>();
        parts.add(header);
        parts.addAll(lines);
        parts.add("---");
        parts.addAll(summary);

        return String.join("\n", parts);
    }

    protected double discount() {
        double luckyDiscountPct = 0.0;
        if (random.nextDouble() < 0.1) {
            luckyDiscountPct = random.nextDouble() * 0.05;
        }
        return luckyDiscountPct;
    }

    protected Date getCurrentDate() {
        return new Date();
    }

    private double roundMoney(double n) {
        return Math.round(n * 100.0) / 100.0;
    }
}
